### PR TITLE
[Coverage] Avoid recording regions from macro expansions

### PR DESCRIPTION
**Pull request by hamishknight**
Ignore any regions recorded while inside a macro expansion, but account for any control flow that may have happened such that the exit count is correctly adjusted. This allows e.g throwing function calls to behave correctly within the expansion.

Additionally, make sure we avoid profiling functions with function body macros.

rdar://129081384

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

File: test/Profiler/coverage_macros.swift [hunk 1]
```diff
@@ -43,7 +70,62 @@ struct S2 {
...
+  guard .random() else { return }
+  print("hello")
+}
+
+// CHECK-LABEL: sil_coverage_map{{.*}}s15coverage_macros5test8yyKF
+func test8() throws {             // CHECK-NEXT: [[@LINE]]:21   -> [[@LINE+4]]:2 : 0
+  guard .random() else { return } // CHECK-NEXT: [[@LINE]]:24   -> [[@LINE]]:34  : 1
+                                  // CHECK-NEXT: [[@LINE-1]]:34 -> [[@LINE+2]]:2 : (0 - 1)
+  #declMacroWithControlFlow       // CHECK-NEXT: [[@LINE]]:28   -> [[@LINE+1]]:2 : ((0 - 1) - 3)
+}
```
Comment (hamishknight) [inline test/Profiler/coverage_macros.swift:131]: Need to also add a case for a nested expansion

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/SIL/IR/SILProfiler.cpp [hunk 2]
```diff
@@ -247,9 +258,10 @@ bool shouldSkipExpr(Expr *E) {
 /// Whether the children of a decl that isn't explicitly handled should be
 /// walked.
 static bool shouldWalkIntoUnhandledDecl(const Decl *D) {
-  // We want to walk into the initializer for a pattern binding decl. This
-  // allows us to map LazyInitializerExprs.
-  return isa<PatternBindingDecl>(D);
+  // We want to walk into initializers for bindings, and the expansions of
+  // MacroExpansionDecls, which will be nested within MacroExpansionExprs in
+  // local contexts.
+  return isa<PatternBindingDecl>(D) || isa<MacroExpansionDecl>(D);
```
Comment (ahoppen) [inline lib/SIL/IR/SILProfiler.cpp:264]: How do we report code coverage for the expansions of a macro? Or I don’t fully understand why we need to walk into `MacroExpansionDecl`.

**Review by ahoppen — APPROVED** [review]
Looks reasonable to me. But I’m not too familiar with code coverage overall.

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/SIL/IR/SILProfiler.cpp (hunk 2 above)
Comment (hamishknight) [reply lib/SIL/IR/SILProfiler.cpp:264]: We don't, the only reason we need to walk into the expansion is to get accurate counter information for the following code, e.g for `#id(try throwingFn())`, we need to account for the fact that code after the macro may not run if an error was thrown

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Pull request merged**